packages/next/shared/lib/router/router.ts

```diff
@@ -2042,73 +2080,67 @@ export default class Router implements BaseRouter {
   }
 }
 
-function matchesMiddleware(params: {
-  fns: [location: string, isSSR: boolean][]
-  asPath: string
-  locale?: string
-}) {
-  const { pathname: asPathname } = parsePath(params.asPath)
-  const cleanedAs = removeLocale(
-    hasBasePath(asPathname) ? removeBasePath(asPathname) : asPathname,
-    params.locale
-  )
-
-  return params.fns.some(([middleware, isSSR]) => {
-    return getRouteMatcher(
-      getMiddlewareRegex(middleware, {
-        catchAll: !isSSR,
-      })
-    )(cleanedAs)
-  })
-}
-
 interface MiddlewareEffectParams<T extends FetchDataOutput> {
-  fetchData: () => Promise<T>
+  fetchData?: () => Promise<T>
   locale?: string
   asPath: string
   router: Router
 }
 
-function withMiddlewareEffects<T extends FetchDataOutput>(
+function matchesMiddleware<T extends FetchDataOutput>(
   options: MiddlewareEffectParams<T>
-) {
+): Promise<boolean> {
   return Promise.resolve(options.router.pageLoader.getMiddlewareList()).then(
     (fns) => {
-      const matches = matchesMiddleware({
-        asPath: options.asPath,
-        locale: options.locale,
-        fns: fns,
-      })
+      const { pathname: asPathname } = parsePath(options.asPath)
+      const cleanedAs = removeLocale(
+        hasBasePath(asPathname) ? removeBasePath(asPathname) : asPathname,
+        options.locale
+      )
 
-      if (matches) {
-        return options
-          .fetchData()
-          .then((data) =>
-            getMiddlewareData(data.dataHref, data.response, options).then(
-              (effect) => ({
-                dataHref: data.dataHref,
-                json: data.json,
-                response: data.response,
-                text: data.text,
-                effect,
-              })
-            )
-          )
-          .catch(() => {
-            /**
-             * TODO: Revisit this in the future.
-             * For now we will not consider middleware data errors to be fatal.
-             * maybe we should revisit in the future.
-             */
-            return null
+      return fns.some(([middleware, isSSR]) => {
+        return getRouteMatcher(
+          getMiddlewareRegex(middleware, {
+            catchAll: !isSSR,
           })
-      }
-
-      return null
+        )(cleanedAs)
+      })
     }
   )
 }
 
+function withMiddlewareEffects<T extends FetchDataOutput>(
+  options: MiddlewareEffectParams<T>
+) {
+  return matchesMiddleware(options).then((matches) => {
+    if (matches && options.fetchData) {
+      return options
+        .fetchData()
+        .then((data) =>
+          getMiddlewareData(data.dataHref, data.response, options).then(
+            (effect) => ({
+              dataHref: data.dataHref,
+              json: data.json,
+              response: data.response,
+              text: data.text,
+              effect,
+            })
+          )
+        )
+        .catch((_err) => {
+          /**
+           * TODO: Revisit this in the future.
```

Comment:
I think we can remove this catch after merging this one https://github.com/vercel/next.js/pull/37550